### PR TITLE
jp/deflation

### PR DESCRIPTION
**Pull request by ioannisPApapadopoulos**
Here are the necessary modifications to the Newton iteration in the `equilibriummeasure` function to utilize deflation when looking for multiple solutions of a one-interval log-kernel equilbrium measures problem. The Newton update is computed as normal and then (nonlinearly) scaled to prevent the Newton method converging to an already discovered solution. 

An example of a problem with 9 solutions (3 feasible) is found in `examples/non-unique-equilibrium-measure.jl`.

### CODE REVIEW

**Comment by TSGut** [comment]
This is really cool! 

In terms of the terminology we've been using, are they all equilibrium measures? I think the global minimizer is called the equilibrium measure and the other potential local minimizers are just referred to as steady states and non uniqueness of steady states is not necessarily the same as non uniqueness of the equilibrium measures. We should maybe compute the energies along with the measured to check this.

**Comment by ioannisPApapadopoulos** [comment]
great, thanks! in terms of the code, does that mean just computing the energies of all the 9 solutions in the example script? Is there anything else you had in mind?

**Comment by dlfivefifty** [comment]
Yes, the energies should be greater than the equilibrium measure which has 3 intervals, and can currently only be computed in Mathematica 

you’ll need the log transform implemented to compute the energy, though point wise evaluation is fine and is easier to compute. Check out ` StieltjesPoint ` in stieltjes.jl for an example with the Cauchy/Stieltjes  kernel

**Comment by ioannisPApapadopoulos** [comment]
Am I wrong in thinking that implementing the log transform for weighted U in this example can be done by simply re-implementing the already coded up log transform for weighted T (stieltjes.jl, lines 157-171)?

**Comment by dlfivefifty** [comment]
https://github.com/JuliaApproximation/ClassicalOrthogonalPolynomials.jl/blob/94d58435970c692060351b68a81d08b41322ff74/src/stieltjes.jl#L157

is only for evaluating the log transform on the same interval. You can tell because `LogKernel{T,R,D}` where `R` is the type of the range interval and `D` is the type of the domain interval, in this case `R == D == ChebyshevInterval`, that is we know the domain and range are the same. 

 What we need for two-intervals is evaluating when `R` is a different interval. 

For comparison, for the Hilbert/Stieltjes case we have `R == D == ChebyshevInterval` here:

https://github.com/JuliaApproximation/ClassicalOrthogonalPolynomials.jl/blob/94d58435970c692060351b68a81d08b41322ff74/src/stieltjes.jl#L87

and `R ≠ D == ChebyshevInterval` here:

https://github.com/JuliaApproximation/ClassicalOrthogonalPolynomials.jl/blob/94d58435970c692060351b68a81d08b41322ff74/src/stieltjes.jl#L256

**Comment by ioannisPApapadopoulos** [comment]
ok! and here we only want to perform the log transform over the range where the (feasible) measures are supported? So not on the whole domain (a,b)?

**Comment by ioannisPApapadopoulos** [comment]
So should the following already be working? 

```
x = 0.1 # some real point, either on the domain or not
W = Weighted(ChebyshevU())
t = axes(W,1)
inv.(x .- t')' * W # should return a row-vector
```
I currently get a `DimensionMismatch` error.

`DimensionMismatch("Second axis of A, Base.OneTo(1), and first axis of B, Inclusion(-1.0..1.0 (Chebyshev)) must match")`

**Comment by dlfivefifty** [comment]
You have an extra ` ‘ `

**Comment by dlfivefifty** [comment]
because I did.. fixed my example

**Comment by ioannisPApapadopoulos** [comment]
without the extra ', I am still getting an error:
```
MethodError: no method matching materialize!(::ArrayLayouts.MulAdd{ArrayLayouts.UnknownLayout, ArrayLayouts.UnknownLayout, LazyArrays.PaddedLayout{ArrayLayouts.DenseColumnMajor}, Float64, QuasiArrays.SubQuasiArray{Float64, 2, QuasiArrays.PermutedDimsQuasiArray{Float64, 2, (2, 1), (2, 1), Weighted{Float64, ChebyshevU{Float64}}}, Tuple{Base.Slice{InfiniteArrays.OneToInf{Int64}}, Inclusion{Float64, ChebyshevInterval{Float64}}}, false}, QuasiArrays.SubQuasiArray{Float64, 1, QuasiArrays.PermutedDimsQuasiArray{Float64, 2, (2, 1), (2, 1), QuasiArrays.BroadcastQuasiMatrix{Float64, typeof(inv), Tuple{QuasiArrays.BroadcastQuasiMatrix{Float64, typeof(-), Tuple{QuasiArrays.QuasiAdjoint{Float64, Inclusion{Float64, ChebyshevInterval{Float64}}}, Inclusion{Float64, ChebyshevInterval{Float64}}}}}}}, Tuple{Inclusion{Float64, ChebyshevInterval{Float64}}, Float64}, false}, LazyArrays.CachedArray{Float64, 1, Vector{Float64}, FillArrays.Zeros{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}}})
Closest candidates are:
  materialize!(::Any, ::Base.Broadcast.Broadcasted{BS, Axes, F, Args} where {Axes, F, Args<:Tuple}) where {Style, BS<:BlockArrays.AbstractBlockStyle} at C:\Users\john.papad\.julia\packages\BlockArrays\rGDmR\src\blockbroadcast.jl:147
  materialize!(::Any, ::Base.Broadcast.Broadcasted{Style, Axes, F, Args} where {Axes, F, Args<:Tuple}) where Style at broadcast.jl:890
  materialize!(::Any, ::Any) at broadcast.jl:886
  ...
```

**Comment by dlfivefifty** [comment]
Ah, for some reason it's only working for complex numbers right now:
```julia
julia> inv.(x .- t') * W
transpose((ComplexF64) .* ((ComplexF64) .^ ℵ₀-element InfiniteArrays.OneToInf{Int64} with indices OneToInf() with indices OneToInf()) with indices OneToInf()) with indices Base.OneTo(1)×OneToInf():
 0.0917376-1.29575im  -0.531751-0.0756742im  -0.0467394+0.21711im  …  
```
I'll have to look into that...

**Comment by dlfivefifty** [comment]
Actually the issue isn't real or complex, it's that the code for evaluation when `t in ChebyshevInterval()` is broken. 

https://github.com/JuliaApproximation/ClassicalOrthogonalPolynomials.jl/pull/66

has a fix.

**Comment by ioannisPApapadopoulos** [comment]
Thanks @dlfivefifty! Switching to the branch `dl/offhilbertcolsupportbug` did the trick and it works on my end now too! 

I suppose that implementing the evaluation of the energy at a point for the log-kernel would be part of a seperate PR for ClassicalOrthogonalPolynomials.jl. Would you want to merge this deflation PR into EquilibriumMeasures.jl or leave it as a branch for now?

**Comment by dlfivefifty** [comment]
Let’s merge this, but perhaps it would be helpful to make it support other types like ` BigFloat ` . Do you want to have a try, or maybe at this point it is more helpful if I go through the code and make the necessary changes so you can learn what to do?

**Comment by ioannisPApapadopoulos** [comment]
I am willing to give it a go! This would be my first time dealing with `BigFloat`, do you have a pointer to somewhere in JuliaApproximation where this has been done before? Maybe a commit where the necessary changes were made? (perhaps I am too optimistic here). Thank you!

**Comment by dlfivefifty** [comment]
This is more basic Julia. Eg ` 0.0 ` vs ` zero(T) `.

some tricks are to use ` @code_warntype ` to decide whether the types are inferred (and so code is fast). For unit tests one uses ` @inferred `

**Comment by ioannisPApapadopoulos** [comment]
Okay! So I think it's working in the sense that giving initial guesses that are `BigFloat` seems to work. Is there something more subtle that I may have missed?

File: src/EquilibriumMeasures.jl [hunk 1]
```diff
@@ -1,122 +1,187 @@
...
+struct EquilibriumMeasureMoment
+    V
+end
+
+function (E::EquilibriumMeasureMoment)(a) 
+    c0,μ = _equilibriummeasure(E.V, a...)
+    SVector(c0, sum(μ) - 1)
+end
+
+function equilibriummeasure(V; a = SVector(-1.0,1.0), maxiterations=1000, knownsolutions=[], power=2, shift=1.0, dampening=1.0, returnEndpoint=false)
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:40]:
```suggestion
function equilibriummeasure(V; a = SVector(-1.0,1.0), maxiterations=1000, knownsolutions=[], power=2, shift=1.0, dampening=1.0, returnendpoint=false)
```
Julia convention is to only uses CamelCase for types. Sometimes `_` are used but tend to be skipped.

File: src/EquilibriumMeasures.jl [hunk 2]
```diff
@@ -1,122 +1,187 @@
...
+        a = an
+    end
+    error("Max its")
+end
+
+# Deflation code
+
+function  deflation_inner_products(state::SVector, sol)
+    num_sols = length(sol)
+    T = promote_type(typeof(state[1]), typeof(sol[1][1]))
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:73]:
```suggestion
    T = promote_type(eltype(state), eltype(eltype(sol)))
```

File: src/EquilibriumMeasures.jl [hunk 3]
```diff
@@ -1,122 +1,187 @@
...
+    inner_products = zeros(T, num_sols)
+    for iter = 1:num_sols
+        inner_products[iter] = dot(state.-sol[iter], state.-sol[iter])
+    end
+    inner_products
+end
+
+function deflation_op(state::SVector, sol, power::Real, shift::Real)
+    num_sols = length(sol)
+    T = promote_type(typeof(state[1]), typeof(sol[1][1]))
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:84]:
```suggestion
    T = promote_type(eltype(state), eltype(eltype(sol)))
```

File: src/EquilibriumMeasures.jl [hunk 4]
```diff
@@ -1,122 +1,187 @@
...
+    power = T(power); shift = T(shift)
+
+    m = one(T)
+    inner_products = deflation_inner_products(state, sol)
+    
+    for iter = 1:num_sols
+        normsq = inner_products[iter]
+        factor = one(T)/normsq^(0.5*power) + shift
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:92]:
```suggestion
        factor = one(T)/normsq^(power/2) + shift
```
`0.5` is a `Float64` so this version is more robust to other types.

File: src/EquilibriumMeasures.jl [hunk 5]
```diff
@@ -1,122 +1,187 @@
...
+        m = m*factor
+    end
+    m, inner_products
+end
+
+function deflation_deriv(update::SVector, state::SVector, sol::Vector, power::Real, shift::Real)
+    m, inner_products = deflation_op(state, sol, power, shift)
+    num_sols = length(sol)
+    T = promote_type(typeof(state[1]), typeof(sol[1][1]))
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:101]:
```suggestion
    T = promote_type(eltype(state), eltype(eltype(sol)))
```

File: src/EquilibriumMeasures.jl [hunk 6]
```diff
@@ -1,122 +1,187 @@
...
+            end
+        end
+        a = an
+    end
+    error("Max its")
+end
+
+# Deflation code
+
+function  deflation_inner_products(state::SVector, sol)
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:71]: Is there a reason to restrict to `SVector`? It seems the code would work for any `AbstractVector`

File: src/EquilibriumMeasures.jl [hunk 7]
```diff
@@ -1,122 +1,187 @@
...
+    m, inner_products = deflation_op(state, sol, power, shift)
+    num_sols = length(sol)
+    T = promote_type(typeof(state[1]), typeof(sol[1][1]))
+    power = T(power); shift = T(shift)
+
+    dm = zero(T)
+    state_dot_update = dot(state, update)
+    for iter = 1:num_sols
+        scale = m/deflation_op(state, sol[iter], power, shift)[1]
+        deriv = -power*state_dot_update/inner_products[iter]^(0.5*(power + 1))
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:108]:
```suggestion
        deriv = -power*state_dot_update/inner_products[iter]^((power + 1)/2)
```

File: src/EquilibriumMeasures.jl [hunk 8]
```diff
@@ -1,122 +1,187 @@
...
+        dm += scale*deriv
+    end
+    m, dm
+end
+
+function deflation_scale(update::SVector, state::SVector, sol::Vector, power::Real, shift::Real)
+    T = promote_type(typeof(state[1]), typeof(sol[1][1]))
```
Comment (dlfivefifty) [inline src/EquilibriumMeasures.jl:115]:
```suggestion
    T = promote_type(eltype(state), eltype(eltype(sol)))
```

**Review by dlfivefifty — CHANGES_REQUESTED** [review]

**Review by dlfivefifty — APPROVED** [review]

**Comment by dlfivefifty** [comment]
For the record: we need the following branch

https://github.com/JuliaDiff/ForwardDiff.jl/pull/541

I'll try to get this working with a custom Manifest.toml

**Pull request merged**